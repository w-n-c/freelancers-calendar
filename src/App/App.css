.App {
	text-align: center;
}

h2 {
	font-size: 1em;
}

table {
	width: 100%;
}

td {
	width: 14.28%;
	width: calc(100% / 7);
}

button {
	box-sizing: border-box;
	border: 0px;
	background: inherit;
	cursor: pointer;
}

a {
	text-decoration: none;
}

ul {
	list-style: none;
	margin: 0;
	padding: 0;
}

header {
	display: flex;
	justify-content: center;
	align-items: baseline;
}

header svg {
	height: 20px;
	width: 30px;
}

nav, nav > ul {
	display: flex;
	align-items: baseline;
}
nav li {
	padding: 8px;
}

/* left float menu button with centered nav */
header button {
	margin-right: auto;
}

header span {
	/* width === (+ widthMenuBtn (- widthTodayLink widthViewBtn ))*/
	width: 39px;
	margin-left: auto;
	visibility: hidden;
}

li {
	display: block;
	float: left;
	padding: 1rem;
	position: relative;
	text-decoration: none;
	transition-duration: 0.5s;
}

li:hover {
	cursor: pointer;
}

.dropdown {
	visibility: hidden;
}

ul li ul {
	min-width: 5rem;
	position: absolute;
	opacity: 0;
	transition: all 0.5s ease;
	left: 0;
	display: none;
}

ul li:hover > ul,
ul li:focus-within > ul {
	visibility: visible;
	opacity: 1;
	display: block;
}

ul li ul li {
	background: white;
	border: 1px solid black;
	box-sizing: border-box;
	clear: both;
	width: 100%;
}
