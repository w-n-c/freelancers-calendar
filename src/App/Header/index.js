import React from 'react'
import { Link } from '@reach/router'
import { pick, curry } from 'lodash/fp'
import {incWeek, decWeek, incMonth, decMonth, getMonthName} from './utils'

const _url = (eventUrl, view, date) => `/${view}/${date}/${eventUrl}`
export const url = curry(_url)

const capitalize = (s) => s.charAt(0).toUpperCase() + s.slice(1);
const viewToNoun = (s) => s.charAt(0).toUpperCase() + s.slice(1,-2);

export default (props) => {
	const { view, today, '*': eventPath, location: { search }} = props
	const day = pick(['year', 'month', 'date'], props)

	const eventInfo = `${eventPath}${search}`
	const eventUrl = url(eventInfo)

	const viewUrl = eventUrl(view)
	const weeklyUrl = eventUrl('weekly')
	const monthlyUrl = eventUrl('monthly')

	const nextMonthUrl = monthlyUrl(incMonth(day))
	const prevMonthUrl = monthlyUrl(decMonth(day))
	const nextWeekUrl  = weeklyUrl(incWeek(day))
	const prevWeekUrl  = weeklyUrl(decWeek(day))
	const todaysUrl    = viewUrl(today)
	const toWeeklyUrl  = weeklyUrl(today)
	const toMonthlyUrl = monthlyUrl(today)

	// had the idea to remove all the ternaries below but need to consider how to deal with the urls also changing
	// probably need to make another component and pase the title and url as props
	// will do next commit
	const title = {
		noun: viewToNoun(view),
		adjective: capitalize(view)
	}

	console.log(title)
	return (
		<header className="site-header">
			<nav role="navigation" aria-labelledby="calendar-date">
				<ul>
					<h1 id="calendar-date">{getMonthName(day.month)} {day.year}</h1>
					<li>
						<Link title="Return to Today" to={todaysUrl}>Today</Link>
					</li>
					<li>
						{props.view === 'monthly' &&
							<Link title="Previous Month" to={prevMonthUrl}>&lt;</Link>
						}{props.view === 'weekly' &&
							<Link title="Previous Week" to={prevWeekUrl}>&lt;</Link>
						}
					</li>
					<li>
						{props.view === 'monthly' &&
							<Link title="Next Month" to={nextMonthUrl}>&gt;</Link>
						}{props.view === 'weekly' &&
							<Link title="Next Week" to={nextWeekUrl}>&gt;</Link>
						}
					</li>
					<li>
						<button aria-haspopup="true">View</button>
						<ul className="dropdown" aria-label="submenu">
							<li>
								<Link title="Weekly View" to={toWeeklyUrl}>
									Weekly
								</Link>
							</li>
							<li>
								<Link title="Monthly View" to={toMonthlyUrl}>
									Monthly
								</Link>
							</li>
						</ul>
					</li>
				</ul>
			</nav>
		</header>
	)
}
