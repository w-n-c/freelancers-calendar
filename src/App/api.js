const uuid = require('nanoid')
const fs = require('fs')

function randInt(max) {
	return Math.floor(Math.random() * Math.floor(max))
}

// Originally ran every time the app started,
// but random events made testing difficult
// now call "node api.js" if events.json doesn't
// exist to create a new static list of events

const createEvents = () => {
	const year = 2019
	const month = 8
	const fakeEvents = []
	let eventNumber = 1
	for (let day = 1; day <= 30; day++) {
		let moreEvents = day % 3
		for(let i = 0; i <= moreEvents; i++) {
			const event = {}
			event.id = uuid()
			event.start = new Date(year, month, day, randInt(24), randInt(60)).toISOString()
			event.end = new Date(new Date(event.start).getTime() + 1000*60*60).toISOString()
			event.title = `event no: ${eventNumber++}`
			event.description = 'description'
			fakeEvents.push(event)
		}
	}
	return fakeEvents
}

let data = createEvents() 
data = JSON.stringify(data)
fs.writeFileSync('events.json', data)
