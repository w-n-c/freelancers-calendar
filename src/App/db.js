import Dexie from 'dexie'

const db = new Dexie('EventsDB')
db.version(1).stores({
	events: '++id',
	eventCache: '++id',
})

export default db
