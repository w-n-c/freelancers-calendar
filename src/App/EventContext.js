import React from 'react'
import { fetchEvents } from './api'

const { Provider, Consumer } = React.createContext()

class EventProvider extends React.Component {
	state = {
		events: [],
		loading: false,
		error: ''
	}

	componentDidMount() {
		this.setState({ loading: true, error: '' })
		fetchEvents()
			.then(events => this.setState({ loading: false, events }))
			.catch(error => this.setState({ loading: false, error}))
	}

	filterTodaysEvents = (year, month, day) => {
		const startDate = new Date(`${year}/${month}/${day}`)
		const endDate = new Date(startDate); endDate.setHours(24)
		const daysEvents = this.state.events.filter(event => {
			return event.start >= startDate && event.start < endDate
		})
		return daysEvents
	}

	// 'delete' > 'destroy' because its the same number of characters as
	// 'create' and 'update' (ocd much?) for this use it means the same thing
	handleCreateEvent(){}
	handleUpdateEvent(){}
	handleDeleteEvent(){}

	render() {
		return (
			<Provider value={{
				...this.state,
				filterTodaysEvents: this.filterTodaysEvents,
				onCreateEvent: this.handleCreateEvent,
				onUpdateEvent: this.handleUpdateEvent,
				onDeleteEvent: this.handleDeleteEvent,
			}}>{this.props.children}</Provider>
		)
	}
}


export { EventProvider, Consumer as EventConsumer }
