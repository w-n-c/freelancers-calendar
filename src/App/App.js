import React, { useState } from 'react'
import { Router, Redirect } from '@reach/router'
import { throttle } from 'lodash'
import './App.css'
import Header from './Header'
import Calendar from './Calendar'
import { EventProvider } from './EventContext'
import { toDateString } from './utils'

export const App = () => {
	const [today, setToday] = useState(toDateString(new Date()))

	const options = {leading: true, trailing: true}
	const handleUpdateToday = throttle(
		() => setToday(toDateString(new Date())),
		1000,
		options
	)
	
	return (
		<EventProvider>
			<Router>
				<Redirect noThrow from="/" to={`monthly/${today}`} />
				<Header path="/:view/:year/:month/:date/*" today={today} />
			</Router>
			<Calendar handleUpdateToday={handleUpdateToday} />
		</EventProvider>
	)
}
