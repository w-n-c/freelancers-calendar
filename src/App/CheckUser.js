import React, { useEffect, useContext } from "react"
import UserContext from './UserContext'
import EventContext from './EventContext'

export default ({navigate}) => {
	const {getUser} = useContext(UserContext)
	useEffect(() => {
		getUser()
		navigate('/')
	})

	return (
		<h1>Loading...</h1>
	)
}
