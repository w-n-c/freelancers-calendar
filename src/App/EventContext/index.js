import EventContext from './EventContext'
export { EventProvider, EventConsumer } from './EventContext'
export default EventContext
