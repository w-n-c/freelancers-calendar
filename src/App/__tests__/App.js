import React from 'react'
import ReactDOM from 'react-dom'
import {App} from '../App'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('<App />', () => {
	it('renders without crashing', () => {
		const div = document.createElement('div')
		ReactDOM.render(<App />, div)
		ReactDOM.unmountComponentAtNode(div)
	})
	it('matches the snapshot', () => {
		const tree = shallow(<App />)
		expect(toJson(tree)).toMatchSnapshot()
	})
})
