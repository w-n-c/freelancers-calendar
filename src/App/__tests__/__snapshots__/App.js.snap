// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches the snapshot 1`] = `
<EventProvider>
  <Router>
    <Redirect
      from="/"
      noThrow={true}
      to="monthly/2019/10/16"
    />
    <_default
      path="/:view/:year/:month/:date/*"
      today="2019/10/16"
    />
  </Router>
  <Calendar
    handleUpdateToday={[Function]}
  />
</EventProvider>
`;
